src/s2_functors.erl

```diff
@@ -0,0 +1,70 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%% @doc The Functor Type Class.
+%%% @end
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+%%%_* Module declaration ===============================================
+-module(s2_functors).
+
+%%%_* Exports ==========================================================
+-export([fmap/2]).
+
+%%%_* Includes =========================================================
+-include("prelude.hrl").
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+%%%_* Types ============================================================
+-type functor(A) :: maybe(A, _)   |
+                    [A]           |
+                    #{_ := A}     |
+                    fun((_) -> A) |
+                    {A}.
+
+%%%_* Code =============================================================
+-spec fmap(fun((A) -> B), functor(A)) -> functor(B).
+%%@doc fmap(F, Functor) is the result of calling F on every value in Functor.
+fmap(F, {ok, _} = Maybe)    when is_function(F, 1)                      -> s2_maybe:fmap(F, Maybe);
+fmap(F, {error, _} = Maybe) when is_function(F, 1)                      -> s2_maybe:fmap(F, Maybe);
+fmap(F, List)               when is_function(F, 1), is_list(List)       -> lists:map(F, List);
+fmap(F, Map)                when is_function(F, 1), is_map(Map)         -> maps:map(fun(_K, V) -> F(V) end, Map);
+fmap(F, Fun)                when is_function(F, 1), is_function(Fun, 1) -> s2_funs:o(F, Fun);
+fmap(F, Tuple)              when is_function(F, 1), is_tuple(Tuple)     -> list_to_tuple(fmap(F, tuple_to_list(Tuple))).
+
+-ifdef(TEST).
+
+%%%_* Tests ============================================================
+%% Add functors to this list to make sure that they satify the functor
+%% laws. The laws should hold true for any `Value`.
+functors(Value) ->
+  [ {ok, Value}
+  , [Value]
+  , #{key => Value}
+  , fun(arg) -> Value end
+  , {Value}].
+
+assertEqual(Expected, Actual) when is_function(Expected), is_function(Actual) ->
+  assertEqual(Expected(arg), Actual(arg));
+assertEqual(Expected, Actual) ->
+  ?assertEqual(Expected, Actual).
+
+preserve_identity_morphism_test() ->
+  CheckLaw = fun(Functor) -> assertEqual(Functor, fmap(fun s2_funs:id/1, Functor)) end,
+  lists:foreach(CheckLaw, functors({ok, 2})).
```

Comment:
You should only pass the value, right?

```suggestion
  lists:foreach(CheckLaw, functors(2)).
```

Reply:
it should be `{ok, 2}`. I wanted `Value` itself to be a `maybe` since the tests should pass for any value of `Value`. using `{ok, 2}` implicitly prevents that somebody can implemented `dict` as a functor. 

the problem is that a list of `maybes` is basically the same thing as a `dict`:

```erlang
[{ok, 3}, {ok, foo}, {error, reason}].
```

If `Value` in the tests above is just `2`, then somebody could try to add an implementation for a `dict` functor at the end:

```erlang
fmap(F, [{K, V} | _Tail] = Dict) -> dict:map(F, Dict); 
```

And this wouldn't clash with the implementation of the `list` functor since the list test would just test `[2]`, and the `dict` test would test `[{key, 2}]`, where `key` is arbitrary. If we test the list functor with a list of maybes `[{ok, 2}]`, then you can only have either a `list` functor or a `dict` functor.

Writing this I realize that I should make this much clearer in the code 😆